### PR TITLE
GH-4985: Transaction vm2 settings

### LINKED ISSUE #4985: [BUG] Put transaction is expecting --transferred-value in VM1 instead of VM2
Identified on the back of the put transaction issue expecting --transferred-value.

Problem with VM-specific fields: Fields like transaction_runtime are being required for all VMs, even though they are only relevant for specific VMs (e.g., VM2). This causes unnecessary complications and creates a risk of technical debt.

Impact:
For VM1, fields like transaction_runtime are irrelevant, leading to confusion and potential runtime errors.
The current structure increases complexity for developers and tools interacting with the transactions.

### PR DESCRIPTION
**Pull request by mpapierski**
Transaction settings valid only for VM2 are stored in an enum. This solves situation when transaction target has specified extra values that will not be used by VM1 runtime.

Closes #4985

### CODE REVIEW

**Comment by EdHastingsCasperAssociation** [comment]
Overall, I think this will work but I'm going to pull this down and make some refinements. Will push one or more commits later today.

**Comment by EdHastingsCasperAssociation** [comment]
See https://github.com/mpapierski/casper-node/pull/9

**Review by EdHastingsCasperAssociation — APPROVED** [review]

File: types/src/gens.rs [hunk 1]
```diff
@@ -1108,60 +1108,80 @@ pub fn transaction_invocation_target_arb() -> impl Strategy<Value = TransactionI
...
+        transaction_stored_runtime_params_arb(),
     )
-        .prop_map(|(target, runtime, transferred_value)| {
-            TransactionTarget::new_stored(target, runtime, transferred_value)
-        })
+        .prop_map(|(id, runtime)| TransactionTarget::Stored { id, runtime })
+}
+
+fn transferred_value_arb() -> impl Strategy<Value = u64> {
+    any::<u64>()
```
Comment (zajko) [inline types/src/gens.rs:1117]: I wonder if having a random value here makes any sense. IMHO we should have [0; u64::MAX] to prove that it handles both ends of the range? maybe MAX/2 if we are super zealous?

File: types/src/transaction/transaction_target.rs [hunk 2]
```diff
@@ -23,6 +21,144 @@ use rand::{Rng, RngCore};
...
+        }
+    }
+}
+
+impl ToBytes for TransactionRuntimeParams {
+    fn to_bytes(&self) -> Result<Vec<u8>, Error> {
+        match self {
+            TransactionRuntimeParams::VmCasperV1 => {
+                CalltableSerializationEnvelopeBuilder::new(vec![
+                    crate::bytesrepr::U8_SERIALIZED_LENGTH,
```
Comment (zajko) [inline types/src/transaction/transaction_target.rs]: I personally prefer to put the vec of fields in a separate function and call it both in to_bytes and serialized_length. That way - if any new fields are introduced - only one place needs updating. You can see that in TransactionV1#serialized_field_lengths

File: types/src/transaction/transaction_target.rs [hunk 3]
```diff
@@ -23,6 +21,144 @@ use rand::{Rng, RngCore};
...
+                transferred_value.serialized_length(),
+                seed.serialized_length(),
+            ]),
+        }
+    }
+}
+
+impl FromBytes for TransactionRuntimeParams {
+    fn from_bytes(bytes: &[u8]) -> Result<(TransactionRuntimeParams, &[u8]), Error> {
+        let (binary_payload, remainder) = CalltableSerializationEnvelope::from_bytes(4, bytes)?;
```
Comment (zajko) [inline types/src/transaction/transaction_target.rs]: The VmCasperV2 variant has 2 fields so - including the enum tag - we should expect at most 3 calltable entries here?

File: types/src/transaction/transaction_target.rs (hunk 2 above)
Comment (mpapierski) [reply types/src/transaction/transaction_target.rs]: Addressed

File: types/src/transaction/transaction_target.rs (hunk 3 above)
Comment (mpapierski) [reply types/src/transaction/transaction_target.rs]: Addressed

**Review by zajko — APPROVED** [review]

File: types/src/gens.rs (hunk 1 above)
Comment (mpapierski) [reply types/src/gens.rs:1117]: I'd keep it consistent with other places where we use `any()` but could limit the values to those that only make sense in given context

**Pull request merged**